File: src/main/java/lb3/part1/FTPClientExample.java
package lb3.part1;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.apache.commons.net.ftp.FTP;
import org.apache.commons.net.ftp.FTPClient;

import java.io.*;

@Slf4j
@RequiredArgsConstructor
public class FTPClientExample {
    private final String server;
    private final int port;
    private final String username;
    private final String password;
    private final static String REMOTE_DIRECTORY = "/ubuntu/dists/";
    private final static String LOCAL_DIRECTORY = "src/main/resources/output";
    private final FTPClient ftpClient = new FTPClient();

    public void start () {
        try {
            // Підключення до FTP серверу
            ftpClient.connect(server, port);
            ftpClient.login(username, password);

            // Перехід до каталогу /ubuntu/dists/
            ftpClient.changeWorkingDirectory(REMOTE_DIRECTORY);

            // Отримання списку назв файлів
            String[] fileNames = ftpClient.listNames();

            // Запис назв файлів до локального файлу
            BufferedWriter writer = new BufferedWriter(new FileWriter(LOCAL_DIRECTORY + "/ubuntu_dists.txt"));
            for (String fileName : fileNames) {
                writer.write(fileName);
                writer.newLine();
                if (fileName.endsWith("updates")){
                    String findingDir = REMOTE_DIRECTORY + fileName + "/main";
                    log.info("Finding manifest in " + findingDir + " ");
                    findAndDownloadManifest(findingDir, LOCAL_DIRECTORY, 1);
                }
            }
            writer.close();
            System.out.println("ubuntu_dists.txt saved");


            System.out.println("Operation completed successfully");
        } catch (IOException e) {
            e.printStackTrace();
            System.out.println("Connection with FTP server failed.");
        }
    }
    private void findAndDownloadManifest(String remoteDirectory, String localDirectory, int iteration) throws IOException {
        // Перехід до вибраної директорії
        ftpClient.changeWorkingDirectory(remoteDirectory);

        // Отримання списку назв файлів у поточній директорії
        String[] fileNames = ftpClient.listNames();
        for (String fileName : fileNames) {
            // Шукати файл MANIFEST
            if (fileName.equalsIgnoreCase("MANIFEST")) {
                System.out.println("MANIFEST founded");
                // Завантаження файлу MANIFEST у відповідну директорію
                String localFilePath = localDirectory + remoteDirectory + "main/" + fileName;

                fileDownloader(localFilePath, fileName);
                System.out.println("MANIFEST downloaded " + localFilePath);
//                log.info("MANIFEST downloaded: " + localFilePath + " ");
                return;
            }
        }

        // Аналізувати вкладені директорії
        String[] subDirectories = ftpClient.listNames();
        for (String subDirectory : subDirectories) {
            if (ftpClient.changeWorkingDirectory(remoteDirectory + "/" + subDirectory)
                    && (subDirectory.startsWith("installer") || iteration != 1)
            ) {
//                System.out.println("finding in " + remoteDirectory + "/" + subDirectory);
                findAndDownloadManifest(remoteDirectory + "/" + subDirectory, localDirectory, iteration + 1);
                ftpClient.changeToParentDirectory();
            }
        }
    }

    private void fileDownloader(String localFilePath, String fileName) throws IOException {
        File localFile = new File(localFilePath);
        localFile.getParentFile().mkdirs(); // Створення відповідних директорій, якщо вони ще не існують

        ftpClient.setFileType(FTP.BINARY_FILE_TYPE);
        BufferedOutputStream outputStream = new BufferedOutputStream(new FileOutputStream(localFile));
        ftpClient.retrieveFile(fileName, outputStream);
        outputStream.close();
    }

    public static void main(String[] args) {
        new FTPClientExample("ftp.ubuntu.com", 21,  "anonymous", "").start();
    }
}

File: src/main/java/lb3/part1/FTPClientExample.js
const ftp = require('basic-ftp');
const fs = require('fs');

const server = 'ftp.ubuntu.com';
const port = 21;
const username = 'anonymous';
const password = '';

const remoteDirectory = '/ubuntu/dists/';
const localDirectory = '/Projects/NP/src/main/resources/output';
async function findAndDownloadManifest(client, remoteDirectory, localDirectory, iteration) {
    // Переход в выбранную директорию
    await client.cd(remoteDirectory);

    // Получение списка файлов в текущей директории
    const fileNames = await client.list();

    for (const fileName of fileNames) {
        // Поиск файла MANIFEST
        if (fileName.name.toLowerCase() === 'manifest') {
            console.log('MANIFEST founded');
            // Загрузка файла MANIFEST в соответствующую директорию
            const localFilePath = `${localDirectory}${remoteDirectory}main/${fileName.name}`;
            await fileDownloader(client, localFilePath, fileName.name);
            console.log(`MANIFEST downloaded ${localFilePath}`);
            return;
        }
    }

    // Анализ вложенных директорий
    const subDirectories = await client.list();
    for (const subDirectory of subDirectories) {
        if (
            (await client.cd(`${remoteDirectory}/${subDirectory.name}`)) &&
            (subDirectory.name.startsWith('installer') || iteration !== 1)
        ) {
            console.log(`finding in ${remoteDirectory}/${subDirectory.name}`);
            await findAndDownloadManifest(client, `${remoteDirectory}/${subDirectory.name}`, localDirectory, iteration + 1);
            await client.cd('..');
        }
    }
}

async function fileDownloader(client, localFilePath, fileName) {
    const localFileDir = localFilePath.substring(0, localFilePath.lastIndexOf('/'));
    fs.mkdirSync(localFileDir, { recursive: true }); // Создание необходимых директорий, если их еще нет

    const outputStream = fs.createWriteStream(localFilePath);
    await client.download(outputStream, fileName);
}

async function start() {
    const client = new ftp.Client();
    try {
        await client.access({
            host: server,
            port: port,
            user: username,
            password: password
        });

        // Переход в каталог /ubuntu/dists/
        await client.cd(remoteDirectory);

        // Получение списка файлов
        const fileNames = await client.list();

        // Запись списка файлов в локальный файл
        const writer = fs.createWriteStream(`${localDirectory}/ubuntu_dists.txt`);
        for (const fileName of fileNames) {
            writer.write(fileName.name);
            writer.write('\n');
            if (fileName.name.endsWith('updates')) {
                const findingDir = `${remoteDirectory}${fileName.name}/main`;
                console.log(`Finding manifest in ${findingDir}`);
                await findAndDownloadManifest(client, findingDir, localDirectory, 1);
            }
        }
        writer.end();
        console.log('ubuntu_dists.txt saved');

        console.log('Operation completed successfully');
    } catch (error) {
        console.error(error);
        console.log('Connection with FTP server failed.');
    } finally {
        client.close();
    }
}

start();
